### PR TITLE
don't block when there are no buffers left

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/DispatchingDisruptor.java [hunk 1]
```diff
@@ -46,7 +47,25 @@ public void close() {
...
+    long backoffMillis = 1;
+    long nextLogTime = 0;
+    while (true) {
+      try {
+        return disruptor.getRingBuffer().tryNext();
+      } catch (InsufficientCapacityException insufficientCapacity) {
+        long now = System.currentTimeMillis();
+        backoffMillis = Math.min(backoffMillis * 2, 1000);
+        if (now > nextLogTime) { // log every 20 seconds
+          log.debug("no buffer available, sleeping for {}ms", backoffMillis);
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/DispatchingDisruptor.java:59]: Definitely good to know. Maybe we should add some metrics for these things as well?

Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/DispatchingDisruptor.java:59]: Agreed, this isn't something I want to merge right now

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Monitor.java [hunk 2]
```diff
@@ -114,6 +116,11 @@ public void onScheduleFlush(final DDAgentWriter agentWriter, final boolean previ
       // not recorded
     }
 
+    @Override
+    public void onBackedUpTraceBuffer() {
+      statsd.incrementCounter("trace.buffer.backlog");
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Monitor.java:121]: 👍

**Comment by dougqh** [comment]
Hmm, yes, we should have thought about the no buffer available case a little more carefully.

This seems like a good step in the right direction.
Although, I keep wondering whether we should switch to something other than Disruptor.  I feel like what we need to do doesn't really fit into Disruptor's model.

**Pull request merged**